compiler/modules/front/expr.bal

```diff
@@ -533,11 +552,102 @@ function codeGenArithmeticBinaryExpr(ExprContext cx, bir:BasicBlock bb, bir:Arit
+        }
     }
-    return cx.unimplementedErr(`not implemented: logical expressions on non-constant values`, pos);
+
+    var evalRightBlock = cx.createBasicBlock();
+    var shortCircuitBlock = cx.createBasicBlock();
+    // JBUG var instead tuple type doesn't work
+    [bir:BasicBlock, bir:BasicBlock] [ifTrue, ifFalse] = isOr ? [shortCircuitBlock, evalRightBlock] : [evalRightBlock, shortCircuitBlock];
+    bir:CondBranchInsn condBranch = { operand: lhs, ifTrue: ifTrue.label, ifFalse: ifFalse.label, pos };
+    block1.insns.push(condBranch);
+    ExprContext rhsCx;
+    if lhsNarrowing != () {
+        if isOr {
+            rhsCx = cx.createNarrowedExprContext(evalRightBlock, lhsNarrowing.binding, lhsNarrowing.ifFalse, lhsNarrowing.basisIfFalse, pos);
+        }
+        else {
+            rhsCx = cx.createNarrowedExprContext(evalRightBlock, lhsNarrowing.binding, lhsNarrowing.ifTrue, lhsNarrowing.basisIfTrue, pos);
+        }
+    }
+    else {
+        rhsCx = cx;
+    }
+    var { result: rhs, block: evalRightBlock2, narrowing: rhsNarrowing } = check codeGenExprForBoolean(rhsCx, evalRightBlock, right);
+    ExprNarrowing? narrowing;
+    if lhsNarrowing == () && rhsNarrowing == () {
+        narrowing = ();
+    }
+    else if isOr {
+        if lhsNarrowing != () && rhsNarrowing != () {
+            if lhsNarrowing.binding.name != rhsNarrowing.binding.name {
+                return cx.unimplementedErr("not implemented: logical expr that narrows multiple variables", pos=pos);
+            }
+            narrowing = {
+                binding: lhsNarrowing.binding,
+                ifTrue: t:union(lhsNarrowing.ifTrue, rhsNarrowing.ifTrue),
+                ifFalse: t:intersect(lhsNarrowing.ifFalse, rhsNarrowing.ifFalse),
+                basisIfTrue: { or: [lhsNarrowing.basisIfTrue, rhsNarrowing.basisIfTrue] },
+                basisIfFalse: { and: [lhsNarrowing.basisIfFalse, rhsNarrowing.basisIfFalse] }
+            };
+        }
+        else {
+            ExprNarrowing n = <ExprNarrowing>(lhsNarrowing ?: rhsNarrowing);
+            narrowing = {
+                binding: n.binding,
+                ifTrue: n.binding.reg.semType, // same as t:union(n.ifTrue, n.binding.reg.semType)
+                ifFalse: n.ifFalse, // same as t:intersect(n.ifFalse, n.binding.reg.semType)
+                basisIfTrue: n.basisIfTrue,
+                basisIfFalse: n.basisIfFalse
+            };
+        }
+    }
+    else {
+        if lhsNarrowing != () && rhsNarrowing != () {
+            if lhsNarrowing.binding.name != rhsNarrowing.binding.name {
+                return cx.unimplementedErr("not implemented: logical expr that narrows multiple variables", pos=pos);
+            }
+            narrowing = {
+                binding: lhsNarrowing.binding,
+                ifTrue: t:intersect(lhsNarrowing.ifTrue, rhsNarrowing.ifTrue),
+                ifFalse: t:union(lhsNarrowing.ifFalse, t:intersect(lhsNarrowing.ifTrue, rhsNarrowing.ifFalse)),
```

Comment:
@jclark 
Spec says this, but `t:union(lhsNarrowing.ifFalse, rhsNarrowing.ifFalse)` also works.

Reply:
![narrowing(1)](https://user-images.githubusercontent.com/1686124/150556026-88e06f89-1aa2-42a4-9114-188b1fbcd934.png)

Reply:
This would cut down the amount of code since or/and becomes dual of each other.